### PR TITLE
fix: Private s3 bucket with CDN & compatible S3 providers

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

⚠️ This is the same work as in https://github.com/strapi/strapi/pull/16256, just that I moved everything to typescript.

When using the private configuration in the s3 provider, we parse assets url to find if they are from the configured provider bucket.

Parsing did not work for some compatible S3 providers ( DO spaces, and others) nor for CDN's.

CDN's are tricky, the ones we have tried (cloudfront) require a different package to sign the urls. The workaround is to keep using the same aws sdk package, and return the signed private bucket url not the signed cdn url.

Returning signed cdn urls would require user customization of the `getSignedUrl` method.

### Why is it needed?

Have a private S3 Provider configuration when using CDN's or other compatible S3 providers 

### How to test it?

- See tests
- Ensure there is nothing broken for image signing

### Related issue(s)/PR(s)
Fixes https://github.com/strapi/strapi/issues/16245
Closes https://github.com/strapi/strapi/pull/16256

### CODE REVIEW

File: packages/providers/upload-aws-s3/src/utils.ts [hunk 1]
```diff
@@ -5,6 +5,29 @@ interface BucketInfo {
   err?: string;
 }
 
+export function isUrlFromBucket(fileUrl: string, bucketName: string, bucketBaseUrl = '') {
```
Comment (nathan-pichon) [inline packages/providers/upload-aws-s3/src/utils.ts:8]: Shouldn't we specify the return type on functions? 🤔

Comment (Marc-Roig) [reply packages/providers/upload-aws-s3/src/utils.ts:8]: do we have a standard for it? I am normally used to let typescript infer the return type.

Comment (Marc-Roig) [reply packages/providers/upload-aws-s3/src/utils.ts:8]: doesn't hurt to add it here I imagine, I will add it!

**Review by nathan-pichon — APPROVED** [review]

File: packages/providers/upload-aws-s3/src/utils.ts (hunk 1 above)
Comment (nathan-pichon) [reply packages/providers/upload-aws-s3/src/utils.ts]: I think it was a subject in the TS workshop we had and we did not manage to a conclusion on that specific point 🤔

**Pull request merged**